src/main/java/picard/sam/markduplicates/UmiGraph.java

```diff
@@ -0,0 +1,211 @@
+ * THE SOFTWARE.
+ */
+
+package picard.sam.markduplicates;
+
+import htsjdk.samtools.DuplicateSet;
+import htsjdk.samtools.SAMRecord;
+import picard.PicardException;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.counting;
+/**
+ * UmiGraph is used to identify UMIs that come from the same original source molecule.  The assumption
+ * is that UMIs with small edit distances are likely to be read errors on the sequencer rather than
+ * distinct molecules.
+ *
+ * The algorithm used here is to join all pairs of UMIs that are within maxEditDistanceToJoin.  It is possible
+ * for a set of UMIs A, B and C to all be considered as part of the same source molecule even if two of the UMIs
+ * have a Hamming distance larger than maxEditDistanceToJoin.  Suppose A = "ATCC", B = "AACC", and C = "AACG"
+ * and maxEditDistanceToJoin = 1.  In this case, A and B are 1 Hamming distance so they are joined, and B and C
+ * are 1 Hamming distance so they are joined.  Because A and B are joined and because B and C are joined, this results
+ * in A and C being joined even though they have a distance of 2.
+ */
+public class UmiGraph {
+    private final List<SAMRecord> records;      // SAMRecords from the original duplicate set considered to break up by UMI
+    private final Map<String, Long> umiCounts;  // Map of UMI sequences and how many times they have been observed
+    private final int[] duplicateSetID;         // ID of the duplicate set that the UMI belongs to, the index is the UMI ID
+    private final String[] umi;                 // Sequence of actual UMI, the index is the UMI ID
+    private final int numUmis;                  // Number of observed UMIs
+    private final String umiTag;                // UMI tag used in the SAM/BAM/CRAM file ie. RX
+    private final String assignedUmiTag;        // Assigned UMI tag used in the SAM/BAM/CRAM file ie. MX
+
+    public UmiGraph(DuplicateSet set, String umiTag, String assignedUmiTag) {
+        this.umiTag = umiTag;
+        this.assignedUmiTag = assignedUmiTag;
+        records = set.getRecords();
+
+        // First ensure that all the reads have a UMI, if any reads are missing a UMI throw an exception
+        for (SAMRecord rec : records) {
+            if(rec.getStringAttribute(umiTag) == null) {
+                rec.setAttribute(umiTag, "");
+            }
+        }
+
+        // Count the number of times each UMI occurs
+        umiCounts = records.stream().collect(Collectors.groupingBy(p -> p.getStringAttribute(umiTag), counting()));
+
+        // At first we consider every UMI as if it were its own duplicate set
+        numUmis = umiCounts.size();
+        umi = new String[numUmis];
+
+        duplicateSetID = new int[numUmis];
+        for (int i = 0; i < numUmis; i++) {
+            duplicateSetID[i] = i;
+        }
+
+        int i = 0;
+        for (Map.Entry<String, Long> entry : umiCounts.entrySet()) {
```

Comment:
I'm getting the keys, not the values, but I see what you mean, and I've changed it so that I get the key directly rather than through the Map.Entry.